openbb-agent-copilot/openbb_agent_copilot/main.py

```diff
@@ -0,0 +1,82 @@
+import re
+import json
+from pathlib import Path
+from typing import AsyncGenerator
+from openbb_agents.agent import openbb_agent
+import os
+
+from fastapi import FastAPI
+from fastapi.middleware.cors import CORSMiddleware
+from fastapi.responses import JSONResponse
+from magentic import (
+    UserMessage,
+    AssistantMessage,
+    AsyncStreamedStr,
+)
+
+from dotenv import load_dotenv
+from .models import AgentQueryRequest
+
+load_dotenv(".env")
+app = FastAPI()
+
+origins = [
+    "http://localhost",
+    "http://localhost:1420",
+    "http://localhost:5050",
+    "https://pro.openbb.dev",
+    "https://pro.openbb.co",
+]
+
+app.add_middleware(
+    CORSMiddleware,
+    allow_origins=origins,
+    allow_credentials=True,
+    allow_methods=["*"],
+    allow_headers=["*"],
+)
+
+
+def sanitize_message(message: str) -> str:
+    """Sanitize a message by escaping forbidden characters."""
+    cleaned_message = re.sub(r"(?<!\{)\{(?!{)", "{{", message)
+    cleaned_message = re.sub(r"(?<!\})\}(?!})", "}}", cleaned_message)
+    return cleaned_message
+
+
+async def create_message_stream(
+    content: AsyncStreamedStr,
+) -> AsyncGenerator[dict, None]:
+    async for chunk in content:
+        yield {"event": "copilotMessageChunk", "data": {"delta": chunk}}
+
+
+@app.get("/copilots.json")
+def get_copilot_description():
+    """Widgets configuration file for the OpenBB Terminal Pro"""
+    return JSONResponse(
+        content=json.load(open((Path(__file__).parent.resolve() / "copilots.json")))
+    )
+
+
+@app.post("/v1/query")
+async def query(request: AgentQueryRequest):
+    """Query the Copilot."""
+
+    chat_messages = []
+    for message in request.messages:
+        if message.role == "ai":
+            chat_messages.append(
+                AssistantMessage(content=sanitize_message(message.content))
+            )
+        elif message.role == "human":
+            chat_messages.append(UserMessage(content=sanitize_message(message.content)))
+
+    try:    
+        result = openbb_agent(str(chat_messages), verbose=False, openbb_pat=os.getenv("OPENBB_PAT"))
+        
+        return {"output": result}
```

Comment:
You'll still need to stream the result back (just as a single chunk) using the `create_message_stream` function. We now use named events for our SSEs for things to be parsed and rendered correctly on the front-end.

Reply:
I actually tried this and had no success with it, can you share if we have more documentation on how to do this? Because in this case the output is literally just a string.

I know we have an async openbb_agent but I looked at the repo, and it's missing the PAT argument which is really important as it grants it access to all my data.

PS: I kept getting that the following screenshot. cc @mnicstruwig 

<img width="575" alt="Screenshot 2024-08-20 at 12 57 51 AM" src="https://github.com/user-attachments/assets/b12a9233-c86f-48f3-a17b-dfb20d3e5b7c">

Reply:
@DidierRLopes I think there might be something I've missed on my part. Let me investigate for you, and I'll give you an update 🙏 .

Reply:
@DidierRLopes Fixed the outputs using SSEs -- we just had to use the `hasStreaming=true` in Copilot (since Terminal Pro officially only supports streaming output now). We just stream back the entire answer in a single chunk.